models/cryptopunks/ethereum/cryptopunks_ethereum_events.sql

```diff
@@ -0,0 +1,124 @@
+{{ config(
+        alias ='events',
+        materialized ='incremental',
+        file_format ='delta',
+        incremental_strategy='merge',
+        unique_key='unique_trade_id'
+        )
+}}
+
+WITH buys AS (
+  SELECT
+    `type`,
+    punkIndex,
+    CASE
+      WHEN buy_value = 0 THEN lag(bid_value) OVER (
+        PARTITION BY punkIndex
+        ORDER BY
+          evt_block_number ASC,
+          evt_index ASC
+      )
+      ELSE buy_value
+    END AS `enhanced_buy_value`,
+    `from`,
+    `bid_from`,
+    CASE
+      WHEN `to` = '0x0000000000000000000000000000000000000000' THEN lag(`bid_from`) OVER (
+        PARTITION BY punkIndex
+        ORDER BY
+          evt_block_number ASC,
+          evt_index ASC
+      )
+      ELSE `to`
+    END AS `enhanced_to`,
+    evt_block_time,
+    evt_block_number,
+    evt_index,
+    evt_tx_hash,
+    contract_address,
+    CASE WHEN (buy_value = 0 AND `to` = '0x0000000000000000000000000000000000000000') THEN "Offer Accepted" ELSE "Buy" END AS trade_category
+  FROM
+    (
+      SELECT
+        "PunkBought" AS `type`,
+        punkIndex,
+        value AS buy_value,
+        NULL AS bid_value,
+        toAddress as `to`,
+        fromAddress AS `from`,
+        NULL AS `bid_from`,
+        evt_block_number,
+        evt_index,
+        evt_block_time,
+        evt_tx_hash,
+        contract_address
+      FROM
+        {{ source('cryptopunks_ethereum','CryptoPunksMarket_evt_PunkBought') }}
+      UNION ALL
+        -- `from` is address that placed the bid
+      SELECT
+        "PunkBidEntered" AS `type`,
+        punkIndex,
+        NULL AS buy_value,
+        value AS bid_value,
+        NULL,
+        NULL AS `from`,
+        fromAddress AS `bid_from`,
+        evt_block_number,
+        evt_index,
+        evt_block_time,
+        evt_tx_hash,
+        contract_address
+      FROM
+        {{ source('cryptopunks_ethereum','CryptoPunksMarket_evt_PunkBidEntered') }}
+    )
+)
+SELECT
+    "ethereum" as blockchain,
+    "cryptopunks" as project,
+    "v1" as `version`,
+    buys.evt_block_time AS block_time,
+    buys.punkIndex as token_id,
+    "CryptoPunks" AS `collection`,
+    buys.enhanced_buy_value / power(10,18) * p.price AS amount_usd,
+    "erc20" AS token_standard,
+    NULL::string AS trade_type,
+    NULL::string AS number_of_items,
+    buys.trade_category,
+    buys.from AS seller,
+    buys.`enhanced_to` AS buyer,
+    "Trade" as evt_type,
```

Comment:
Since all evt_types in this spell are actually trades, there are two options:
Either rename this spell a cryptopunks_ethereum_trades.sql

Or, if we want to keep the possibility of having different types of events for cryptopunks, we should build the cryptopunks_ethereum_trades model anyways, using something like 
here: https://github.com/duneanalytics/spellbook/blob/main/models/looksrare/ethereum/looksrare_ethereum_trades.sql

Reply:
👍 integrating trades model in next commit